### PR TITLE
gh-144278: Enable overriding TAG (sys.implementation.cache_tag) when building sysmodule.c

### PR DESCRIPTION
**Pull request by zooba**
In particular, this allows defining in to NULL, which has the effect of disabling automatic .pyc caching.
We also update tests to handle the cache tag being None, which was always allowed, but was not accounted for in tests.

* Issue: gh-144278

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -26,6 +26,7 @@ concurrency:
 
 env:
   FORCE_COLOR: 1
+  PIP_NO_COMPILE: 1
```
Comment (StanFromIreland) [inline .github/workflows/build.yml]: Env vars defined with `env:` at the workflow level are not automatically inherited by reusable workflows, hence the CI failures. Currently we simply just define them in each one, I suggest you just do that for `PIP_NO_COMPILE`.

We could pass them as inputs, but for two variables it is not that important IMO, and it can always be done in a follow up.

Comment (zooba) [reply .github/workflows/build.yml]: Yeah, I've just figured that out. I'm not planning on leaving the NULL `cache_tag` in there by default, so it won't matter once that's undone, but will be good to get 100% pass before I turn it off again.

File: Lib/compileall.py [hunk 2]
```diff
@@ -165,6 +165,14 @@ def compile_file(fullname, ddir=None, force=False, rx=None, quiet=0,
     stripdir = os.fspath(stripdir) if stripdir is not None else None
     name = os.path.basename(fullname)
 
+    # Without a cache_tag, we can only create legacy .pyc files. None of our
+    # callers seem to expect this, so the best we can do is fail without raising
+    if not legacy and sys.implementation.cache_tag is None:
+        if not quiet:
+            print("No cache tag is available to generate .pyc path for {!r}"
+                    .format(fullname))
```
Comment (webknjaz) [inline Lib/compileall.py]: Any reason not to use an f-string?
```suggestion
            print(f"No cache tag is available to generate .pyc path for {fullname!r}")
```

Comment (zooba) [reply Lib/compileall.py]: Partly consistency with the code around it, partly because that puts it over the line length limit and so now it has to wrap in an awkward way. But no real strong reason to do anything at all - the best option is probably just to `, repr(fullname)`

**Comment by zooba** [comment]
If anyone stumbles into here and has any idea why execution is 2-3x slower when `cache_tag` is None, I'd love to hear about it. The test failures/cancellations are timeouts/time-sensitivity issues, not poor functionality, but I can't come up with any reason why it would be so much slower.

File: Lib/test/support/import_helper.py [hunk 3]
```diff
@@ -49,20 +50,31 @@ def forget(modname):
...
     :return: The file system path to the legacy pyc file.
     """
-    pyc_file = importlib.util.cache_from_source(source)
     assert source.endswith('.py')
     legacy_pyc = source + 'c'
-    shutil.move(pyc_file, legacy_pyc)
+    try:
+        pyc_file = importlib.util.cache_from_source(source)
+        shutil.move(pyc_file, legacy_pyc)
+    except (FileNotFoundError, NotImplementedError):
```
Comment (brettcannon) [inline Lib/test/support/import_helper.py:74]:
```suggestion
    except FileNotFoundError, NotImplementedError:
```

File: Lib/zipfile/__init__.py [hunk 4]
```diff
@@ -2260,101 +2259,58 @@ def writepy(self, pathname, basename="", filterfunc=None):
...
-            import py_compile
-            if self.debug:
-                print("Compiling", file)
-            try:
-                py_compile.compile(file, doraise=True, optimize=optimize)
-            except py_compile.PyCompileError as err:
-                print(err.msg)
-                return False
-            return True
+        import importlib._bootstrap_external
```
Comment (brettcannon) [inline Lib/zipfile/__init__.py:2282]: I look forward to PEP 810 landing.

**Review by brettcannon — APPROVED** [review]
I only reviewed the non-test files.
